### PR TITLE
Support cast vote by signature in ZrxTreasury

### PR DESCRIPTION
**Pull request by cece-z**
## Description

Support cast vote by signature in ZrxTreasury

## Checklist:

-   [x] Prefix PR title with `[WIP]` if necessary.
-   [x] Add tests to cover changes as needed.
-   [x] Update documentation as needed.
-   [x] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/treasury/contracts/src/ZrxTreasury.sol [hunk 1]
```diff
@@ -165,43 +174,38 @@ contract ZrxTreasury is
...
+        bool support,
+        bytes32[] memory operatedPoolIds,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    )
+        public
+        override
+    {
+        bytes32 domainSeparator = keccak256(abi.encode(DOMAIN_TYPEHASH, keccak256(bytes(name)), _getChainId(), address(this)));
```
Comment (Recmo) [inline contracts/treasury/contracts/src/ZrxTreasury.sol:202]: `domainnSeparator` is basically constant and a good candidate for an [`immutable`](https://docs.soliditylang.org/en/v0.8.6/contracts.html?highlight=immutable#immutable) set once in the constructor.

Comment (Recmo) [reply contracts/treasury/contracts/src/ZrxTreasury.sol:202]: This is how it is done elsewhere in the codebase: https://github.com/0xProject/protocol/blob/24397c51a8c7bf704948c8fc6874843bccd5d244/contracts/zero-ex/contracts/src/fixins/FixinEIP712.sol#L35

Comment (cece-z) [reply contracts/treasury/contracts/src/ZrxTreasury.sol:202]: Should I directly use the domain separator from `FixinEIP712`? @Recmo

File: contracts/test-utils/src/vote_factory.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            return !_.isString(value) ? value.toString() : value;
+        });
+        return eip712Utils.createTypedData(
+            // TODO(need to bump @0x/json-schemas first?)
+            constants.ZEROEX_VOTE_SCHEMA.name,
+            { ZeroExVote: constants.ZEROEX_VOTE_SCHEMA.parameters },
+            normalizedVote,
+            this._domain,
+        );
+    }
```
Comment (cece-z) [inline contracts/test-utils/src/vote_factory.ts:68]: Hi @Recmo , can you help confirm that for this test, I will need to define/update in 0x/tools/json-schemas for this? 🙏

Comment (mzhu25) [reply contracts/test-utils/src/vote_factory.ts:68]: I'd say let's move this stuff to `@0x/contracts-treasury`. There's a chance the website will use this tooling, so best to keep everything together I think.
Also I personally prefer the abstractions used [here](https://github.com/0xProject/protocol/blob/development/packages/protocol-utils/src/orders.ts) for EIP712 stuff.

File: contracts/test-utils/src/types.ts [hunk 3]
```diff
@@ -201,3 +201,13 @@ export enum ExchangeFunctionName {
...
 }
+
+export interface ZeroExVote {
+    proposalId: BigNumber;
+    support: boolean;
+    operatedPoolIds: number[];
+}
+
+export interface SignedZeroExVote extends ZeroExVote {
+    signature: string;
```
Comment (mzhu25) [inline contracts/test-utils/src/types.ts:212]:
```suggestion
    v: number;
    r: string;
    s: string;
```

File: contracts/test-utils/src/vote_factory.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        this._delegatorPrivateKey = delegatorPrivateKey;
+        this._domain = {name: contractName, chainId, verifyingContract};
+    }
+
+    public async newSignedZeroExVoteAsync(
+        vote: ZeroExVote,
+        signatureType: SignatureType = SignatureType.EIP712,
+    ): Promise<SignedZeroExVote> {
+        const voteHashBuffer = this._getVoteHashBuffer(vote);
+        const signature = signingUtils.signMessage(voteHashBuffer, this._delegatorPrivateKey, signatureType);
```
Comment (mzhu25) [inline contracts/test-utils/src/vote_factory.ts:44]: You can use [this function](https://github.com/0xProject/protocol/blob/development/packages/protocol-utils/src/signature_utils.ts#L112) to sign, that way you won't need to export `parseSignatureHexAsRSV`

**Review by mzhu25 — COMMENTED** [review]
Great work! Thanks for taking on this task 😄  
Solidity looks pretty much good to me, left some suggestions on the TypeScript.

File: contracts/treasury/src/votes.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+                EIP712Domain: EIP712_DOMAIN_PARAMETERS,
+                [Vote.STRUCT_NAME]: Vote.STRUCT_ABI,
+            },
+            domain: Vote._createVoteEIP712Domain(this.chainId, this.verifyingContract) as any,
+            primaryType: Vote.STRUCT_NAME,
+            message: {
+                proposalId: this.proposalId.toString(10),
+                support: this.support ? 1 : 0,
+                // TODO(Cece): key types does not support array
+                operatedPoolIds: this.operatedPoolIds,
```
Comment (cece-z) [inline contracts/treasury/src/votes.ts:98]: @mzhu25 thanks for the review and suggestions. Please lmk if this file makes sense to you. I also got this problem that `EIP712TypedData` does not support array as key type, any suggestion how can I resolve this? 🙏  cc: @Recmo

File: contracts/test-utils/src/vote_factory.ts (hunk 2 above)
Comment (cece-z) [reply contracts/test-utils/src/vote_factory.ts]: Thanks Michael, I have moved these to `contracts/treasury/src/votes.ts`, not really confident about the code structure, would be great to get your opinion on that. thx!

File: contracts/treasury/contracts/src/ZrxTreasury.sol (hunk 1 above)
Comment (Recmo) [reply contracts/treasury/contracts/src/ZrxTreasury.sol]: No, it's better to create unique domain separators for each distinct use of signatures. The domain separators are there to make sure a signature is not accidentally valid for two different use-cases.

File: contracts/treasury/contracts/src/ZrxTreasury.sol [hunk 6]
```diff
@@ -165,43 +187,38 @@ contract ZrxTreasury is
...
+    ///      See `getVotingPower` for how voting power is computed.
+    /// @param proposalId The ID of the proposal to vote on.
+    /// @param support Whether to support the proposal or not.
+    /// @param operatedPoolIds The pools operated by `msg.sender`. The
+    ///        ZRX currently delegated to those pools will be accounted
+    ///        for in the voting power.
+    /// @param v the v field of the signature
+    /// @param r the r field of the signature
+    /// @param s the s field of the signature
+    function castVoteBySignature(
```
[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply contracts/treasury/contracts/src/ZrxTreasury.sol:204]: Idk maybe something like n * 700 + 2.6k (eip 2929) extra. Not that bad.

File: contracts/test-utils/src/types.ts [hunk 7]
```diff
@@ -161,3 +161,17 @@ export enum ExchangeFunctionName {
...
+    operatedPoolIds: string[];
+    chainId: number;
+    verifyingContract: string;
+}
+
+export interface ZeroExSignedVote extends ZeroExVote {
+    v: number;
+    r: string;
+    s: string;
+}
```
[the remark this answers is not in the document]
Comment (cece-z) [reply contracts/test-utils/src/types.ts]: yah you are right, I added them at the beginning of working on this change, followed some other examples. I dont need this anymore, cleaned up.

File: contracts/treasury/contracts/src/IZrxTreasury.sol [hunk 8]
```diff
@@ -59,12 +59,36 @@ interface IZrxTreasury {
 
     event VoteCast(
         address voter,
+        address msgSender,
```
[the remark this answers is not in the document]
Comment (cece-z) [reply contracts/treasury/contracts/src/IZrxTreasury.sol]: yah make sense, removed. thx

File: contracts/treasury/contracts/src/IZrxTreasury.sol [hunk 9]
```diff
@@ -150,6 +174,27 @@ interface IZrxTreasury {
...
+    ///      Only callable during the voting period for that proposal.
+    ///      See `getVotingPower` for how voting power is computed.
+    /// @param proposalId The ID of the proposal to vote on.
+    /// @param support Whether to support the proposal or not.
+    /// @param operatedPoolIds The pools operated by the signer. The
+    ///        ZRX currently delegated to those pools will be accounted
+    ///        for in the voting power.
+    /// @param v the v field of the signature
+    /// @param r the r field of the signature
+    /// @param s the s field of the signature
```
[the remark this answers is not in the document]
Comment (cece-z) [reply contracts/treasury/contracts/src/IZrxTreasury.sol:165]: updated for both castVote and castVoteBySignature. thx

File: contracts/treasury/src/votes.ts [hunk 10]
```diff
@@ -0,0 +1,109 @@
+import { ZeroExSignedVote } from '@0x/contracts-test-utils/src/types';
```
[the remark this answers is not in the document]
Comment (cece-z) [reply contracts/treasury/src/votes.ts]: Got rid of the types and removed this import. thx

File: contracts/treasury/src/votes.ts [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+            hexUtils.toHex(
+                hexUtils.concat(
+                    '0x1901',
+                    this.getDomainHash(),
+                    this.getStructHash()),
+            ),
+        );
+    }
+
+    public getSignatureWithKey(privateKey: string): ZeroExSignedVote {
```
[the remark this answers is not in the document]
Comment (cece-z) [reply contracts/treasury/src/votes.ts]: got it, updated. thx!

File: contracts/treasury/contracts/src/ZrxTreasury.sol [hunk 12]
```diff
@@ -373,4 +390,64 @@ contract ZrxTreasury is
...
+        }
+
+        if (support) {
+            proposals[proposalId].votesFor = proposals[proposalId].votesFor
+            .safeAdd(votingPower);
+        } else {
+            proposals[proposalId].votesAgainst = proposals[proposalId].votesAgainst
+            .safeAdd(votingPower);
+        }
+        hasVoted[proposalId][voter] = true;
```
[the remark this answers is not in the document]
Comment (mzhu25) [reply contracts/treasury/contracts/src/ZrxTreasury.sol:435]: `getVotingPower` is `view`, but yes hypothetically if it made a non-static external call there would be a reentrancy bug

File: packages/protocol-utils/src/votes.ts [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    support: false,
+    operatedPoolIds: [] as string[],
+    chainId: 1,
+    version: '1.0.0',
+    verifyingContract: NULL_ADDRESS,
+};
+
+export type VoteFields = typeof VOTE_DEFAULT_VALUES;
+
+export class Vote {
```
Comment (dorothy-zbornak) [inline packages/protocol-utils/src/votes.ts:19]: Can we rename this to `TreasuryVote`?

File: packages/protocol-utils/src/votes.ts [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+const VOTE_DEFAULT_VALUES = {
+    proposalId: ZERO,
+    support: false,
+    operatedPoolIds: [] as string[],
+    chainId: 1,
+    version: '1.0.0',
+    verifyingContract: NULL_ADDRESS,
+};
+
+export type VoteFields = typeof VOTE_DEFAULT_VALUES;
```
Comment (dorothy-zbornak) [inline packages/protocol-utils/src/votes.ts:17]: And `TreasuryVoteFields`

File: contracts/treasury/test/treasury_test.ts [hunk 15]
```diff
@@ -465,6 +472,110 @@ blockchainTests.resets('Treasury governance', env => {
...
         });
+        // castVoteBySignature()
+        it('Cannot vote by signature on invalid proposalId', async () => {
+            await fastForwardToNextEpochAsync();
+            await fastForwardToNextEpochAsync();
+            const vote = new Vote({
+                proposalId: INVALID_PROPOSAL_ID,
+                verifyingContract: admin,
+            });
+            const signedVote = vote.getSignatureWithKey(delegatorPrivateKey);
```
Comment (dorothy-zbornak) [inline contracts/treasury/test/treasury_test.ts:483]:
```suggestion
            const signature = vote.getSignatureWithKey(delegatorPrivateKey);
```
etc

File: contracts/treasury/test/treasury_test.ts [hunk 16]
```diff
@@ -17,8 +17,9 @@ import {
     randomAddress,
     verifyEventsFromLogs,
 } from '@0x/contracts-test-utils';
-import { BigNumber } from '@0x/utils';
-import * as _ from 'lodash';
+import { Vote } from '@0x/protocol-utils/lib/src/votes';
```
Comment (dorothy-zbornak) [inline contracts/treasury/test/treasury_test.ts:20]: So what you need to do is export `Vote` from `src/index.ts` in `protocol-utils` so you can just do
```suggestion
import { Vote } from '@0x/protocol-utils';
```

**Review by dorothy-zbornak — COMMENTED** [review]
The end is in sight!

Don't forget to update `CHANGELOG.json` in both `treasury` and `protocol-utils` so the publish job knows that these packages need to be republished ([example](https://github.com/0xProject/protocol/pull/285/files#diff-8846f1e484b9312305932169b6d7348d034a072b7d8e28a58cb607ea4ba73cba)). Here are the basic rules:

- If the first version object (most recent) has a `timestamp`, it has already been published so you have to prepend a new version object, with your PR as the only item in in the `changes` array, bumping up the minor version number by 1. Do not add a `timestamp` field.
- If the  first entry does not have a `timestamp` field, it hasn't been published yet so you can just append your PR to its `changes` array.

**Review by dorothy-zbornak — APPROVED** [review]
Epic. Thank you!

**Review by mzhu25 — APPROVED** [review]
Awesome work! 🎉

**Pull request merged**